woocommerce/woocommerce
# Display "Origin" column in Orders table in Orders Analytics

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/46383.

In this PR: 

- We add a new "Origin" column into the Orders table in Orders Analytics. 
    - The `GET /wp-json/wc-analytics/reports/orders` endpoint is modified to return the origin data in `extended_info.attribution.origin` in the order objects in the API response. Example screenshot below:

<img width="367" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/635780ac-3c96-4231-92a6-aee77dfeccf5">

- We add a new "Origin" column into the exported CSV file when users click on the "Download" button in the Orders table in Orders Analytics.
    - Note that there are two methods of generating the CSV file export: client side CSV generation when there are 25 or less records, and server side CSV generation when there are more than 25 records. See https://github.com/woocommerce/woocommerce/blob/9f9d0268cd6f601b9c3d8bdd51ddf93b11a5df41/docs/reporting/adding-columns-to-analytics-reports-and-csv-downloads.md for more info.

Screenshot below shows the new "Origin" column in the Orders table in Orders Analytics:

<img width="322" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/4904fb67-0a2b-46b8-a462-1ccddc663fdf">

Screenshot below shows the new "Origin" column in the exported CSV file:

<img width="242" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/9b1de587-e54a-4ebd-820a-29b47c8328b3">




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

NOTE: Orders Analytics has a cache in place. To test the code, you may refer to commit https://github.com/woocommerce/woocommerce/pull/46424/commits/1a48761a0cfd215bca68839ce8f1261fe2ad65d6 to disable caching.

1. Create some orders with different order attribution origin. You may refer to PR https://github.com/woocommerce/woocommerce/pull/44907 on how to do that. 
    - You should have less than 25 orders to test client side CSV export file generation.
    - Alternatively, you can also use the date filter in Orders Analytics page to adjust the number of orders to be displayed.
3. Go to Orders Analytics: `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
4. In the Orders table at the bottom of the page, there should be a new "Origin" column. Make sure the values are correct for the orders, i.e. they should match what you see in the Orders listing page or in Order details page.
5. In the table header, click on the "Download" button. A CSV file should be generated and downloaded immediately. Check that the CSV file contains the "Origin" column, and the values are correct for the orders.
6. Create more orders with different order attribution origin. You should have more than 25 orders to test server side CSV export file generation.
7. Go to Orders Analytics: `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
8. In the table header, click on the "Download" button. You should see a notification toast at the bottom that says you will receive an email with a link to download the CSV file. You may need to wait for a few seconds or minutes to get the email. In the email, click on the link to get the CSV file. Check that the CSV file contains the "Origin" column, and the values are correct for the orders.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
